### PR TITLE
[BEAM-744] A runner should be able to override KafkaIO max wait prope…

### PR DESCRIPTION
**Pull request by amitsela**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

…rties.

Add KafkaOptions for the UnboundedKafkaReader.

### CODE REVIEW

**Comment by amitsela** [comment]
R: @rangadi and @dhalperi (committer).
CC: @aljoscha, @tgroh (Flink/Direct runners).

**Comment by amitsela** [comment]
Jenkins failures seem to be related to `Dataflow` IT tests - `NoSuchMethodError` - could it be a dirty classpath ?

**Comment by dhalperi** [comment]
I think that this is generally on the wrong path. Runners should not need to override temporal constants in specific transforms to get sane behavior. I believe the simple rule of thumb should be "readers should return as soon as they are able" + "runners may poll advance() in a loop for a certain period of time if it returned too fast" + "runners must tolerate sources that take a long time to start or advance, because real systems operate that way".

I think we're violating all of these in various places, but that combined these principles add up to a good solution. Thoughts?

**Comment by amitsela** [comment]
I'd be happy to summarize this once we have something, and I agree with what you write Dan, but it seems that this was an issue for the `DirectRunner` as explained in this [conversation](https://github.com/apache/incubator-beam/pull/1071).
According to your suggestion, we should set the two "wait" properties to something like 10 msec, correct ?
If the `DirectRunner` can handle it, I'd be happy to change the ticket, PR, and commit.

**Comment by dhalperi** [comment]
per conversation with @tgroh , I believe we can and should do this now.

**Comment by tgroh** [comment]
The `DirectRunner` should reuse readers for some amount of time, and forever until they produce elements. #1128 fixes the current state to actually continue to read from a shard that ever returned false in an initial evaluation, but is not directly related to this change.

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 1]
```diff
@@ -757,7 +757,7 @@ public void validate() {
 
     private static final Duration KAFKA_POLL_TIMEOUT = Duration.millis(1000);
     // how long to wait for new records from kafka consumer inside start()
-    private static final Duration START_NEW_RECORDS_POLL_TIMEOUT = Duration.standardSeconds(5);
+    private static final Duration START_NEW_RECORDS_POLL_TIMEOUT = Duration.millis(10);
```
Comment (rangadi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: Better to remove this then.

Recently when I modified KafkaIOTest, I removed a bit of extra code that handled 'false' from start(). I need to put that back. I can send a separate PR for that.

**Comment by rangadi** [comment]
>  I believe the simple rule of thumb should be "readers should return as soon as they are able" + "runners may poll advance() in a loop for a certain period of time if it returned too fast" + "runners must tolerate sources that take a long time to start or advance, because real systems operate that way".

I like these. :+1:

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java (hunk 1 above)
Comment (amitsela) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: So we'll only have `NEW_RECORDS_POLL_TIMEOUT`, sure why not.

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 2]
```diff
@@ -756,9 +756,7 @@ public void validate() {
     private Iterator<PartitionState> curBatch = Collections.emptyIterator();
 
     private static final Duration KAFKA_POLL_TIMEOUT = Duration.millis(1000);
-    // how long to wait for new records from kafka consumer inside start()
-    private static final Duration START_NEW_RECORDS_POLL_TIMEOUT = Duration.standardSeconds(5);
-    // how long to wait for new records from kafka consumer inside advance()
+    // how long to wait for new records from kafka consumer.
```
Comment (rangadi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: Can you add 'inside advance()' or 'inside advance()/start()' to this comment? Would make it more clear where the time out is.

**Comment by rangadi** [comment]
Fixed KafkaIOTest in #1133

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 3]
```diff
@@ -968,7 +966,7 @@ public void run() {
 
       // Wait for longer than normal when fetching a batch to improve chances a record is available
       // when start() returns.
-      nextBatch(START_NEW_RECORDS_POLL_TIMEOUT);
```
Comment (rangadi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: actually, can you remove this arg for nextBatch and use `NEW_RECORDS_POLL_TIMEOUT` diretly inside nextBatch().

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 4]
```diff
@@ -968,7 +966,7 @@ public void run() {
 
       // Wait for longer than normal when fetching a batch to improve chances a record is available
```
Comment (rangadi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: remove this comment.

**Comment by amitsela** [comment]
@rangadi I've addressed your comments, PTAL.

File: sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java [hunk 5]
```diff
@@ -888,13 +885,14 @@ private void consumerPollLoop() {
...
+    private void nextBatch() {
       curBatch = Collections.emptyIterator();
 
       ConsumerRecords<byte[], byte[]> records;
       try {
-        records = availableRecordsQueue.poll(timeout.getMillis(),
-                                             TimeUnit.MILLISECONDS);
+        // poll available records, wait (if necessary) up to the specified timeout.
+        records = availableRecordsQueue.poll(NEW_RECORDS_POLL_TIMEOUT.getMillis(),
+            TimeUnit.MILLISECONDS);
```
Comment (rangadi) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: optional/minor : align the args?

Comment (amitsela) [inline sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java]: I really don't mind but it seems like we don't have a consensus on arg-alignment..
I'll align and commit, thanks!

**Pull request closed without merging**